pkg/controller/cluster/agent/shared.go

```diff
@@ -0,0 +1,222 @@
+package agent
+
+import (
+	"fmt"
+
+	"github.com/rancher/k3k/pkg/apis/k3k.io/v1alpha1"
+	"github.com/rancher/k3k/pkg/controller/util"
+	apps "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	virtualKubeletImage = "rancher/k3k:k3k-kubelet"
+)
+
+type SharedAgent struct {
+	cluster   *v1alpha1.Cluster
+	serviceIP string
+}
+
+func NewSharedAgent(cluster *v1alpha1.Cluster, serviceIP string) Agent {
+	return &SharedAgent{
+		cluster:   cluster,
+		serviceIP: serviceIP,
+	}
+}
+
+func (s *SharedAgent) Config() (ctrlruntimeclient.Object, error) {
+	config := sharedAgentData(s.cluster)
+
+	return &v1.Secret{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Secret",
+			APIVersion: "v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      util.AgentConfigName(s.cluster),
+			Namespace: util.ClusterNamespace(s.cluster),
+		},
+		Data: map[string][]byte{
+			"config.yaml": []byte(config),
+		},
+	}, nil
+}
+
+func sharedAgentData(cluster *v1alpha1.Cluster) string {
+	return fmt.Sprintf(`clusterName: %s
+clusterNamespace: %s
+nodeName: %s
+token: %s`, cluster.Name, cluster.Namespace, cluster.Name+"-"+"k3k-kubelet", cluster.Spec.Token)
+}
+
+func (s *SharedAgent) Resources() ([]ctrlruntimeclient.Object, error) {
+	var objs []ctrlruntimeclient.Object
+	objs = append(objs, s.serviceAccount(), s.role(), s.roleBinding(), s.deployment())
+	return objs, nil
+}
+
+func (s *SharedAgent) deployment() *apps.Deployment {
+	selector := metav1.LabelSelector{
+		MatchLabels: map[string]string{
+			"cluster": s.cluster.Name,
+			"type":    "agent",
+			"mode":    "shared",
+		},
+	}
+	name := s.cluster.Name + "-" + "k3k-kubelet"
+	return &apps.Deployment{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Deployment",
+			APIVersion: "apps/v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: util.ClusterNamespace(s.cluster),
+			Labels:    selector.MatchLabels,
+		},
+		Spec: apps.DeploymentSpec{
+			Selector: &selector,
+			Template: v1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: selector.MatchLabels,
+				},
+				Spec: s.podSpec(virtualKubeletImage, name, &selector),
+			},
+		},
+	}
+}
+
+func (s *SharedAgent) podSpec(image, name string, affinitySelector *metav1.LabelSelector) v1.PodSpec {
+	args := []string{"--config", "/opt/rancher/k3k/config.yaml"}
```

Comment:
Is this path guaranteed to exist?

Reply:
Yes so the default path is /etc/rancher/k3k/config.yaml however later in the pod spec you can see that we volume mount the config at /opt/rancher/k3k/config.yaml, but I can make it a constant since its used in two places